certora/specs/Reverts.spec

```
@@ -0,0 +1,283 @@
+    }
+}
+
+invariant preLCFIncreasing()
+    currentContract.PRE_LCF_1 <= currentContract.PRE_LCF_2
+    && currentContract.PRE_LCF_1 <= WAD()
+{
+    preserved {
+        requireInvariant preLIFNotZero();
+    }
+}
+
+invariant preLIFNotZero()
+    0 < currentContract.PRE_LIF_1;
+
+invariant preLIFIncreasing()
+    WAD() < currentContract.PRE_LIF_1
+    && currentContract.PRE_LIF_1 <= currentContract.PRE_LIF_2
+    && currentContract.PRE_LIF_2 <= wDD(WAD(),currentContract.LLTV)
+{
+    preserved {
+        requireInvariant lltvNotZero();
+    }
+}
+
+
+// Check that preLiquidate reverts when its inputs are not validated.
+rule preLiquidateInputValidation(env e, address borrower, uint256 seizedAssets, uint256 repaidShares, bytes data) {
+    requireInvariant preLltvLTlltv();
+    requireInvariant preLCFIncreasing();
+    requireInvariant preLIFIncreasing();
+    preLiquidate@withrevert(e, borrower, seizedAssets, repaidShares, data);
+    assert !exactlyOneZero(seizedAssets, repaidShares) => lastReverted;
+}
+
+// Check that collateralQuoted == 0 would revert by failing require-statements.
+rule zeroCollateralQuotedReverts() {
+    // Market values.
+    uint256 mTotalBorrowAssets;
+    uint256 mTotalBorrowShares;
+
+    // Position values.
+    uint256 pBorrowShares;
+    uint256 pCollateral;
+
+    uint256 collateralPrice;
+
+    requireInvariant preLltvLTlltv();
+
+    uint256 collateralQuoted = require_uint256(summaryMulDivDown(pCollateral, collateralPrice, ORACLE_SCALE()));
+    uint256 borrowed = require_uint256(tAU(pBorrowShares, mTotalBorrowAssets, mTotalBorrowShares));
+
+    uint256 higherBound = wMD(collateralQuoted, currentContract.LLTV);
+    uint256 lowerBound = wMD(collateralQuoted, currentContract.PRE_LLTV);
+
+    assert  collateralQuoted == 0 => (lowerBound >= borrowed || borrowed > higherBound);
+}
+
+// Check that a liquidatable position implies that ltv > PRE_LLTV holds.
+rule preLiquidatableEquivlLtvLTPreLltv() {
+    // Market values.
+    uint256 mTotalBorrowAssets;
+    uint256 mTotalBorrowShares;
+
+    // Position values.
+    uint256 pBorrowShares;
+    uint256 pCollateral;
+
+    uint256 collateralPrice;
+
+    requireInvariant preLltvLTlltv();
+
+    uint256 collateralQuoted = require_uint256(summaryMulDivDown(pCollateral, collateralPrice, ORACLE_SCALE()));
+
+    // Safe require because the implementation would revert, see rule zeroCollateralQuotedReverts.
+    require collateralQuoted > 0;
+
+    uint256 borrowed = require_uint256(tAU(pBorrowShares, mTotalBorrowAssets, mTotalBorrowShares));
+    uint256 ltv = require_uint256(wDU(borrowed,collateralQuoted));
+    uint256 lowerBound = wMD(collateralQuoted, currentContract.PRE_LLTV);
+
+    assert (lowerBound < borrowed) <=> ltv > currentContract.PRE_LLTV;
+}
+
+// Check that pre-liqudidating a position such that ltv <= PRE_LLTV would revert.
+rule nonLiquidatablePositionReverts(env e,address borrower, uint256 seizedAssets, bytes data) {
```

couldn't we prove this also
```suggestion
rule liquidatablePositionReverts(env e,address borrower, uint256 seizedAssets, bytes data) {
```
that's even more trivial (same computation) but good to have it

---

Implemented in https://github.com/morpho-org/pre-liquidation/pull/37/commits/4b1dc0a6ac503942c35ac02ba1cccc52990fea4d.